### PR TITLE
[SystemZ] Add custom handling of legal vectors with reduce-add.

### PR DESCRIPTION
**Pull request by dominik-steenken**
This commit skips the expansion of the `vector.reduce.add` intrinsic on vector-enabled SystemZ targets in order to introduce custom handling of `vector.reduce.add` for legal vector types using the VSUM instructions. This is limited to full vectors with scalar types up to `i32` due to performance concerns.

It also adds testing for the generation of such custom handling, and adapts the related cost computation, as well as the testing for that.

The expected result is a performance boost in certain benchmarks that make heavy use of `vector.reduce.add` with other benchmarks remaining constant.

For instance, the assembly for `vector.reduce.add<4 x i32>` changes from
```hlasm
        vmrlg   %v0, %v24, %v24
        vaf     %v0, %v24, %v0
        vrepf   %v1, %v0, 1
        vaf     %v0, %v0, %v1
        vlgvf   %r2, %v0, 0
```
to
```hlasm
        vgbm    %v0, 0
        vsumqf  %v0, %v24, %v0
        vlgvf   %r2, %v0, 3
```

### CODE REVIEW

**Comment by dominik-steenken** [comment]
@uweigand FYI

File: llvm/lib/Target/SystemZ/SystemZISelLowering.cpp [hunk 1]
```diff
@@ -444,6 +446,11 @@ SystemZTargetLowering::SystemZTargetLowering(const TargetMachine &TM,
       setOperationAction(ISD::SRL, VT, Custom);
       setOperationAction(ISD::ROTL, VT, Custom);
 
+      // Add ISD::VECREDUCE_ADD as custom in order to implement
+      // it with VZERO+VSUM
+      if (Subtarget.hasVector()) {
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:451]: I don't think we need this check - this whole block is executed only if the vector type is "legal", which pre-supposes "hasVector".

File: llvm/lib/Target/SystemZ/SystemZISelLowering.cpp [hunk 2]
```diff
@@ -9505,3 +9514,39 @@ SDValue SystemZTargetLowering::lowerGET_ROUNDING(SDValue Op,
...
+
+  // load a 0 vector for the third operand of VSUM.
+  SDValue Zero = DAG.getSplatBuildVector(OpVT, DL, DAG.getConstant(0, DL, VT));
+
+  // execute VSUM.
+  switch (OpVT.getScalarSizeInBits()) {
+  case 8:
+  case 16:
+    Op = DAG.getNode(SystemZISD::VSUM, DL, MVT::v4i32, Op,
+                     DAG.getBitcast(OpVT, Zero));
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:9536]: Doesn't Zero already have the correct type here, so we don't need the bitcast?

File: llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp [hunk 3]
```diff
@@ -1326,3 +1323,31 @@ SystemZTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
...
+  auto *Type = cast<FixedVectorType>(II->getOperand(0)->getType());
+  unsigned NumElts = Type->getNumElements();
+  unsigned ScalarSize = Type->getScalarSizeInBits();
+  unsigned MaxElts = SystemZ::VectorBits / ScalarSize;
+  if (NumElts < MaxElts)
+    return true;
+
+  // Otherwise
+  switch (II->getIntrinsicID()) {
+  // Do not expand vector.reduce.add
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp:1342]: Comment seems wrong?

File: llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp [hunk 4]
```diff
@@ -1326,3 +1323,31 @@ SystemZTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
...
+  case Intrinsic::vector_reduce_add:
+    // Except for i64, since the performance benefit is dubious there
+    if (ScalarSize < 64) {
+      return false;
+    } else {
+      return true;
+    }
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp:1349]: No braces around single-line statements.  Also, this whole test is maybe simpler as `return ScalarSize >= 64;` ?

File: llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp [hunk 5]
```diff
@@ -1326,3 +1323,31 @@ SystemZTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
...
+  default:
+    return true;
+  }
+}
```
Comment (uweigand) [inline llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp:1353]: Missing newline at the end of the file?

**Review by uweigand — COMMENTED** [review]
Generally looks good, but see a couple of inline comments.

File: llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp (hunk 3 above)
Comment (uweigand) [reply llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp:1342]: Ah sorry, I misread.  Comment is actually correct.

File: llvm/lib/Target/SystemZ/SystemZISelLowering.cpp (hunk 1 above)
Comment (dominik-steenken) [reply llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:451]: ok, i will remove the check.

File: llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp (hunk 4 above)
Comment (dominik-steenken) [reply llvm/lib/Target/SystemZ/SystemZTargetTransformInfo.cpp:1349]: Agreed. that's definitely clearer.

File: llvm/lib/Target/SystemZ/SystemZISelLowering.cpp (hunk 2 above)
Comment (dominik-steenken) [reply llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:9536]: yes, for 8 and 16 bit scalars we do not need the bitcast. I will remove it.

**Comment by dominik-steenken** [comment]
@uweigand Thank you for the comments, I believe i have implemented them all.

**Review by uweigand — APPROVED** [review]
LGTM now, thanks!

**Pull request merged**